Modules/Core/Common/include/itkImageSink.hxx

```diff
@@ -0,0 +1,176 @@
+  InputImageRegionType inputImageRegion = inputPtr->GetLargestPossibleRegion();
+
+  return this->GetRegionSplitter()->GetNumberOfSplits( inputImageRegion, this->m_NumberOfStreamDivisions );
+}
+
+template <class TInputImage>
+void
+ImageSink<TInputImage>
+::GenerateNthInputRequestedRegion (unsigned int inputRequestedRegionNumber)
+{
+  Superclass::GenerateInputRequestedRegion();
+
+  InputImageType* inputPtr = const_cast< InputImageType * >( this->GetInput() );
+  InputImageRegionType inputImageRegion = inputPtr->GetLargestPossibleRegion();
+
+
+  this->GetRegionSplitter()->GetSplit( inputRequestedRegionNumber,
+                                       this->GetNumberOfInputRequestedRegions( ),
+                                       inputImageRegion );
+  m_CurrentInputRegion = inputImageRegion;
+
+  itkDebugMacro( "Generating " << inputRequestedRegionNumber << " chunk as " <<  m_CurrentInputRegion );
+
+  for ( unsigned int idx = 0; idx < this->GetNumberOfInputs(); ++idx )
+    {
+    if ( this->GetInput(idx) )
+      {
+      // Check whether the input is an image of the appropriate
+      // dimension (use ProcessObject's version of the GetInput()
+      // method sinke it returns the input as a pointer to a
+      // DataObject as opposed to the subclass version which
+      // static_casts the input to an TInputImage).
+      using ImageBaseType = ImageBase< InputImageDimension >;
+      typename ImageBaseType::ConstPointer constInput =
+        dynamic_cast< ImageBaseType const * >( this->ProcessObject::GetInput(idx) );
+
+      // If not an image, skip it, and let a subclass of
+      // ImageToImageFilter handle this input.
+      if ( constInput.IsNull() )
+        {
+        continue;
+        }
+
+      // Input is an image, cast away the constness so we can set
+      // the requested region.
+      InputImagePointer input =
+        const_cast< TInputImage * >( this->GetInput(idx) );
+
+      // copy the requested region of the first input to the others
+      InputImageRegionType inputRegion;
+      input->SetRequestedRegion( m_CurrentInputRegion );
+      }
+    }
+}
+
+
+template <class TInputImage>
+void
+ImageSink<TInputImage>
+::StreamedGenerateData( unsigned int itkNotUsed(inputRequestedRegionNumber) )
+{
+
+  this->GetMultiThreader()->SetNumberOfWorkUnits( this->GetNumberOfWorkUnits() );
+  const ThreadIdType  total = this->GetRegionSplitter()->GetNumberOfSplits( m_CurrentInputRegion,  this->GetNumberOfWorkUnits() );
+
+  this->GetMultiThreader()->ParallelizeArray(
+    0,
+    total,
+    [this, total](SizeValueType threadId)
+    {
+        InputImageRegionType splitRegion = this->m_CurrentInputRegion;
+        this->GetRegionSplitter()->GetSplit( threadId, total, splitRegion );
+        this->ThreadedStreamedGenerateData(splitRegion );
+    },
+    nullptr); // \Modules\Core\Common\include\itkProgressTransformer.h would need to be used to properly report progress here
```

Comment:
This comment was more for you @blowekamp. Adding `ProgressTransformer` should be only few extra lines of code, and would increase granularity of progress reporting.

Reply:
See line ~85 of `StreamingProceessObject` where the progress is reported just at the per streaming chunk level. Do you have some example code which would be similar to this situation with 2 levels of progress?

Reply:
It should be similar to this:
https://github.com/InsightSoftwareConsortium/ITK/blob/91ab86f63eb5fc9799bb25fa49e76adf0d1b5251/Modules/Filtering/DisplacementField/include/itkInvertDisplacementFieldImageFilter.hxx#L136-L137

Reply:
But then the other progress reporting is redundant.

Reply:
Thanks for the link.